src/metatrain/experimental/pet/utils/systems_to_batch_dict.py

```diff
@@ -182,6 +198,90 @@ def remap_to_contiguous_indexing(
 
 
 def get_system_batch_dict(
+    system: System,
+    options: NeighborListOptions,
+    all_species: torch.Tensor,
+    max_num_neighbors: torch.Tensor,
+    selected_atoms_index: torch.Tensor,
+    device: torch.device,
+    debug: bool = False,
+) -> Dict[str, torch.Tensor]:
+    if debug:
+        write_system_data(system, options, selected_atoms_index)
+    nl = system.get_neighbor_list(options)
+    i_list = nl.samples.column("first_atom")
+    j_list = nl.samples.column("second_atom")
+
+    unique_neighbors_index, _ = torch.unique(i_list, return_counts=True)
+    unique_index = torch.unique(
+        torch.cat((selected_atoms_index, unique_neighbors_index))
+    )
+
+    # We calculate the actual size of the system, which is the number of
+    # unique atoms in the system.
+    # This is required for LAMMPS interface, because by default
+    # it produces the system with both local and ghost atoms.
+    actual_system_size = len(unique_index)
+
+    # Then we remap the indices of the atoms to a contiguous format.
+    # Also see the docstring of the function for more details.
+    i_list, j_list, unique_neighbors_index = remap_to_contiguous_indexing(
+        i_list, j_list, unique_neighbors_index, unique_index, device
+    )
+
+    index = torch.argsort(i_list, stable=True)
+    j_list = j_list[index]
+    i_list = i_list[index]
+    S_list: torch.Tensor = (
+        torch.cat(
+            (
+                nl.samples.column("cell_shift_a")[None],
+                nl.samples.column("cell_shift_b")[None],
+                nl.samples.column("cell_shift_c")[None],
+            )
+        )
+        .transpose(0, 1)[index]
+        .to(torch.int64)
+    )
+
+    D_list: torch.Tensor = nl.values[:, :, 0][index]
+
+    species = system.types[unique_index].to(torch.int64)
+
+    res = process_neighbors(
+        i_list,
+        j_list,
+        S_list,
+        D_list,
+        max_num_neighbors,
+        torch.tensor(actual_system_size),
+        species,
+        all_species,
+    )
+
+    neighbors_index = res[0]
+    relative_positions = res[1]
+    nums = res[2]
+    mask = res[3]
+    neighbor_species = res[4]
+    neighbors_pos = res[5]
+    species_mapped = res[6]
+
+    system_dict = {
+        "central_species": species_mapped,
+        "x": relative_positions,
+        "neighbor_species": neighbor_species,
+        "neighbors_pos": neighbors_pos,
+        "neighbors_index": neighbors_index,
+        "nums": nums,
+        "mask": mask,
+    }
+    if debug:
+        write_batch_dict(system_dict)
+    return system_dict
+
+
+def get_system_batch_dict_old(
```

Comment:
is this function still used?

Reply:
It is not used, but this is a pure torch implementation of the converter, that (1) contains the code that we can use for benchmarking and (2) contains a lot of comments on how the data is stored in PET. In principle, I can remove it, but I want to at least save the comments somehow.

Reply:
you can add a large comment on top of this file explaining the data format!